store/mem.go

```diff
@@ -2,89 +2,129 @@ package store
+		if err != nil {
+			// The interface for `NewMemStore` is part of the public interface so
+			// adding an error return would be a breaking change so we panic instead.
+			// As of this writing, `lru.New` can only return an error if you pass
+			// maxKeys <= 0 so this should never occur.
+			panic(err)
+		}
+
+		m = &MemStore{
+			keys: keys,
 		}
 	} else {
-		m = &memStore{
-			m: make(map[string]*counter),
+		m = &MemStore{
+			m: make(map[string]*int64),
 		}
 	}
 	return m
 }
 
-// A counter represents a single entry in the MemStore.
-type counter struct {
-	n  int
-	ts time.Time
+// GetWithTime returns the value of the key if it is in the store or
+// -1 if it does not exist. It also returns the current local time on
+// the machine.
+func (ms *MemStore) GetWithTime(key string) (int64, time.Time, error) {
+	now := time.Now()
+	valP, ok := ms.get(key, false)
+
+	if !ok {
+		return -1, now, nil
+	}
+
+	return atomic.LoadInt64(valP), now, nil
 }
 
-// Incr increments the counter for the specified key. It returns the new
-// count value and the remaining number of seconds, or an error.
-func (ms *memStore) Incr(key string, window time.Duration) (int, int, error) {
+// SetIfNotExistsWithTTL sets the value of key only if it is not
+// already set in the store it returns whether a new value was set. It
+// ignores the ttl.
+func (ms *MemStore) SetIfNotExistsWithTTL(key string, value int64, _ time.Duration) (bool, error) {
+	_, ok := ms.get(key, false)
+
+	if ok {
+		return false, nil
+	}
+
 	ms.Lock()
 	defer ms.Unlock()
-	var c *counter
-	if ms.keys != nil {
-		v, _ := ms.keys.Get(key)
-		if v != nil {
-			c = v.(*counter)
-		}
-	} else {
-		c = ms.m[key]
-	}
-	if c == nil {
-		c = &counter{0, time.Now().UTC()}
+
+	_, ok = ms.get(key, true)
+
+	if ok {
+		return false, nil
 	}
-	c.n++
+
+	// Store a pointer to a new instance so that the caller
+	// can't mutate the value after setting
+	v := value
+
 	if ms.keys != nil {
-		ms.keys.Add(key, c)
+		ms.keys.Add(key, &v)
 	} else {
-		ms.m[key] = c
+		ms.m[key] = &v
 	}
-	return c.n, throttled.RemainingSeconds(c.ts, window), nil
+
+	return true, nil
 }
 
-// Reset resets the counter for the specified key. It sets the count
-// to 1 and initializes the timestamp with the current time, in UTC.
-// It returns an error if the operation fails.
-func (ms *memStore) Reset(key string, win time.Duration) error {
-	ms.Lock()
-	defer ms.Unlock()
-	c := &counter{1, time.Now().UTC()}
+// CompareAndSwapWithTTL atomically compares the value at key to the
+// old value. If it matches, it sets it to the new value and returns
+// true. Otherwise, it returns false. If the key does not exist in the
+// store, it returns false with no error. It ignores the ttl.
+func (ms *MemStore) CompareAndSwapWithTTL(key string, old, new int64, _ time.Duration) (bool, error) {
+	valP, ok := ms.get(key, false)
+
+	if !ok {
+		return false, nil
+	}
+
+	return atomic.CompareAndSwapInt64(valP, old, new), nil
```

Comment:
Admittedly the godocs basically say "don't use this unless you're really sure you know what you're doing". I could use the mutex here instead if you'd prefer the clarity.